VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/Util/TextureDecoding.cpp

```diff
@@ -80,28 +80,37 @@ FTextureData DecodeTexture(UObject* Outer, const FString& Key, const FString& Pa
 {
 	EPixelFormat PixelFormat = TextureMetadata.PixelFormat;
 
-	if (TextureMetadata.PixelFormat == EPixelFormat::PF_R8G8B8A8)
+	// Workaround: Also convert grayscale images to rgba, since texture params don't automatically update their sample method
+	if (TextureMetadata.PixelFormat == EPixelFormat::PF_R8G8B8A8 ||
+	    TextureMetadata.PixelFormat == EPixelFormat::PF_G8 ||
+	    TextureMetadata.PixelFormat == EPixelFormat::PF_G16)
 	{
+		const size_t BytesPerBand = TextureMetadata.BytesPerBand;
+		PixelFormat = (BytesPerBand == 2) ? EPixelFormat::PF_A16B16G16R16 : EPixelFormat::PF_B8G8R8A8;
+		const bool bIsColor = (TextureMetadata.Bands >= 3);
+
 		size_t NewBufferSize = TextureMetadata.Width * TextureMetadata.Height * 4 * TextureMetadata.BytesPerBand;
 		auto NewBuffer = std::make_unique<uint8_t[]>(NewBufferSize);
 
 		for (int Y = 0; Y < TextureMetadata.Height; ++Y)
 		{
 			for (int X = 0; X < TextureMetadata.Width; ++X)
 			{
-				const int NewOffset = (Y * TextureMetadata.Width + X) * 4;
-				const int OldOffset = ((TextureMetadata.Height - Y - 1) * TextureMetadata.Width + X) * TextureMetadata.Bands;
-				NewBuffer[NewOffset + 0] = Buffer[OldOffset + 2];
-				NewBuffer[NewOffset + 1] = Buffer[OldOffset + 1];
-				NewBuffer[NewOffset + 2] = Buffer[OldOffset + 0];
-				NewBuffer[NewOffset + 3] = TextureMetadata.Bands == 4 ? Buffer[OldOffset + 3] : 0;
+				const int NewOffset = (Y * TextureMetadata.Width + X) * 4 * BytesPerBand;
+				const int OldOffset = ((TextureMetadata.Height - Y - 1) * TextureMetadata.Width + X) * TextureMetadata.Bands * BytesPerBand;
+				for (int B = 0; B < BytesPerBand; ++B)
+				{
+					NewBuffer[NewOffset + 0 * BytesPerBand + B] = Buffer[OldOffset + 2 * bIsColor + B];
+					NewBuffer[NewOffset + 1 * BytesPerBand + B] = Buffer[OldOffset + 1 * bIsColor + B];
+					NewBuffer[NewOffset + 2 * BytesPerBand + B] = Buffer[OldOffset + 0 * bIsColor + B];
+					NewBuffer[NewOffset + 3 * BytesPerBand + B] = (TextureMetadata.Bands == 4) ? Buffer[OldOffset + 3 * bIsColor + B] : 0;

```

Comment:
I would prefer a ternary here instead of multiplying with the bool ;) Up to you.

Reply:
fixed